apps/prairielearn/src/pages/instructorCourseAdminInstances/instructorCourseAdminInstances.html.ts

```diff
@@ -50,77 +51,134 @@ export function InstructorCourseAdminInstances({
+                              <td class="align-left">${row.formatted_start_date}</td>
+                              <td class="align-left">${row.formatted_end_date}</td>
+                              <td class="align-middle">${row.enrollment_count}</td>
+                            </tr>
+                          `;
+                        })}
+                      </tbody>
+                    </table>
+                  </div>
+                `
+              : html`
+                  <div class="my-4 card-body text-center" style="text-wrap: balance;">
+                    <p class="font-weight-bold">No course instances found.</p>
+                    <p class="mb-0">
+                      A course instance is a single offering of a course, such as "Fall 2024" or
+                      "Fall 2024, Section M".
+                    </p>
+                    <p>
+                      Learn more in the
+                      <a
+                        href="https://prairielearn.readthedocs.io/en/latest/courseInstance/"
+                        target="_blank"
+                        >course instance documentation</a
+                      >.
+                    </p>
+                    ${resLocals.authz_data.has_course_permission_edit &&
+                    !resLocals.course.example_course &&
+                    !resLocals.needToSync
+                      ? html`
+                          <form class="ml-auto" name="add-course-instance-form" method="POST">
+                            <input
+                              type="hidden"
+                              name="__csrf_token"
+                              value="${resLocals.__csrf_token}"
+                            />
+                            <button
+                              name="__action"
+                              value="add_course_instance"
+                              class="btn btn-sm btn-primary"
+                            >
+                              <i class="fa fa-plus" aria-hidden="true"></i>
+                              <span>Add course instance</span>
+                            </button>
+                          </form>
+                        `
+                      : ''}
+                    ${!resLocals.authz_data.has_course_permission_edit
+                      ? html` <p>Course Editors can create new course instances.</p> `
+                      : ''}
+                    ${resLocals.course.example_course
+                      ? html` <p>You can't add course instances to the Example Course.</p> `
+                      : ''}
+                    ${resLocals.needToSync
+                      ? html`
+                          <p>
+                            Sync the course from the <strong>Sync</strong> page before creating a
+                            new course instance.
+                          </p>
+                        `
+                      : ''}
```

Comment:
This leaves us with the interesting possibility of showing multiple messages. I'd be inclined to show at most one, with the following precedence:

- Handle the example course first (permissions / syncing won't change if a course can be added)
- Handle the permissions next (one can't sync the course without editor permissions)
- Handle the "need to sync" case last

We have a tiny helper function (https://github.com/PrairieLearn/PrairieLearn/tree/master/packages/run) that can make this kind of thing nice to write, if you like it:

```ts
${run(() => {
  if (resLocals.course.example_course) {
    return html`<p>You can't add course instances to the Example Course.</p>`;
  }

  if (!resLocals.authz_data.has_course_permission_edit) {
    return html`<p>Course Editors can create new course instances.</p>`;
  }

  if (resLocals.needToSync) {
    return html`
      <p>
        Sync the course from the <strong>Sync</strong> page before creating a new course instance.
      </p>
    `;
  }

  return html`
    <form class="ml-auto" name="add-course-instance-form" method="POST">
      <input type="hidden" name="__csrf_token" value="${resLocals.__csrf_token}" />
      <button name="__action" value="add_course_instance" class="btn btn-sm btn-primary">
        <i class="fa fa-plus" aria-hidden="true"></i>
        <span>Add course instance</span>
      </button>
    </form>
  `;
})}
```

I opted to merge this with the code that determines when to show the button; this ensures that we always show either the button or exactly one message. Let me know if you like that.

Reply:
Thanks for showing me the run function, @nwalters512 . I added this change.